S_find_locale_from_environment: Handle disparate LC_ALL

khwilliamson opened the pull request:
This PR was split off of https://github.com/Perl/perl5/pull/21340.   There are comments in that that apply to this.

When calling setlocale(foo, ""), there is a standardized result.  If there is an environment variable LC_ALL set, it is used regardless of what 'foo' is.

In general, LC_ALL may be set to a heterogenous set of categories and locales, so this could be equivalently

 setlocale(LC_TIME, "de_DE/C/en_US/zh_TW.UTF-8/he_IL");

Prior to this commit, the code did not handle this possibility. It turns out that on all boxes and shells we've tried, they prohibit the LC_ALL environment variable from being set to anything but a single locale name.  But it's easy to handle this eventuality with an already-existing function whose return must be examined and processed.  And doing so makes this code less prone to future changes in those boxes.

jkeenan commented:
> This PR was split off of #21340. There are comments in that that apply to this.
> 
> When calling setlocale(foo, ""), there is a standardized result. If there is an environment variable LC_ALL set, it is used regardless of what 'foo' is.
> 
> In general, LC_ALL may be set to a heterogenous set of categories and locales, so this could be equivalently
> 
> setlocale(LC_TIME, "de_DE/C/en_US/zh_TW.UTF-8/he_IL");
> 
> Prior to this commit, the code did not handle this possibility. It turns out that on all boxes and shells we've tried, they prohibit the LC_ALL environment variable from being set to anything but a single locale name. But it's easy to handle this eventuality with an already-existing function whose return must be examined and processed. And doing so makes this code less prone to future changes in those boxes.

Can we get a unit test(s) that will exercise the various blocks in this p.r.?

khwilliamson committed:
t/run/locale.t: Change var name and hoist to larger scope

The old name was somewhat misleading

khwilliamson committed:
t/run/locale.t: Extract code into a sub

This is in preparation for it to be called from another place

khwilliamson committed:
t/run/locale.t: Save sub return in an array

This is in preparation for it to be used again.

khwilliamson committed:
S_find_locale_from_environment: Handle disparate LC_ALL

When calling setlocale(foo, ""), there is a standardized result.  If
there is an environment variable LC_ALL set, it is used regardless of
what 'foo' is.

In general, LC_ALL may be set to a heterogenous set of categories and
locales, so this could be equivalently

 setlocale(LC_TIME, "de_DE/C/en_US/zh_TW.UTF-8/he_IL");

Prior to this commit, the code did not handle this possibility.  But
it's easy to handle this eventuality by calling an already-existing
function and examining its return.

However, the accompanying test fails at least on Linux, showing that
libc also doesn't handle this possibility.

jkeenan commented:
> > This PR was split off of #21340. There are comments in that that apply to this.
> > When calling setlocale(foo, ""), there is a standardized result. If there is an environment variable LC_ALL set, it is used regardless of what 'foo' is.
> > In general, LC_ALL may be set to a heterogenous set of categories and locales, so this could be equivalently
> > setlocale(LC_TIME, "de_DE/C/en_US/zh_TW.UTF-8/he_IL");
> > Prior to this commit, the code did not handle this possibility. It turns out that on all boxes and shells we've tried, they prohibit the LC_ALL environment variable from being set to anything but a single locale name. But it's easy to handle this eventuality with an already-existing function whose return must be examined and processed. And doing so makes this code less prone to future changes in those boxes.
> 
> Can we get a unit test(s) that will exercise the various blocks in this p.r.?

@khwilliamson, ^^.

khwilliamson commented:
I figured out how to do a test, and found that glibc on Linux did not handle the case properly.  Therefore I'm closing this request, and will add a comment in the appropriate place in locale.c in case we ever want to revisit this.

khwilliamson commented:
A patch file containing this is attached
[disparate.txt](https://github.com/user-attachments/files/16826296/disparate.txt)
